SumoLogic.Logging.Log4Net/BufferedSumoLogicAppender.cs

```diff
@@ -325,17 +325,17 @@ protected override void OnClose()
         {
             base.OnClose();
 
-            if (this.SumoLogicMessageSender != null)
-            {
-                this.SumoLogicMessageSender.Dispose();
-                this.SumoLogicMessageSender = null;
-            }
-
             if (this.flushBufferTimer != null)
             {
                 this.flushBufferTimer.Dispose();
                 this.flushBufferTimer = null;
             }
+
+            if (this.SumoLogicMessageSender != null)
+            {
+                this.SumoLogicMessageSender.Dispose();
+                this.SumoLogicMessageSender = null;
+            }

```

Comment:
Not an expert in log4net, but it has a dedicated `Flush`-method, so one doesn't need to also do it in `OnClose`. Only reason why I changed the `OnClose`-method was to close the `Timer`-first. But I can revert this change if it offends your eye.

Reply:
Regarding the disposing of timer in `ActivateOptions` then it most likely a failsafe, because one is not completely sure of the lifetime of log4net-appender-objects. Still not an expert in log4net, so I will probably stay away from disturbing old giants unless being the author of log4net-library.

Reply:
No your code doesn't call OnClose() or Flush(), but the log4net-framework make sure to call those methods. You just need to implement the contract. Perform close-logic in OnClose() and perform flush-logic in Flush(). No expert in log4net, but pretty sure it flushes and then closes/disposes its appenders on shutdown (without you needing to call anything).

Reply:
If it makes the code-review easier for you, then I can revert my changes for the log4net-appender.

Reply:
No, it isn't a problem reviewing it, I just want to make sure we are not missing anything. 
 
I also feel that `Flush` and `Close` should be getting called together in `Log4Net`, but sadly, I still can't find `Flush` getting called either in 
https://github.com/apache/logging-log4net/blob/master/src/Repository/Hierarchy/Logger.cs
or 
https://github.com/apache/logging-log4net/blob/master/src/Appender/AppenderSkeleton.cs

The destructor of `AppenderSkeleton` seems to call `Close` -> `OnClose`. But no `Flush`.

The one place where I found `Flush` getting called is in `BufferingAppenderSkeleton.OnClose` (https://github.com/apache/logging-log4net/blob/master/src/Appender/BufferingAppenderSkeleton.cs) 
I wonder if `BufferedSumoLogicAppender` had to extend `BufferingAppenderSkeleton` instead of `AppenderSkeleton`.

Reference for all possible usages of `Flush` keyword in their repo: https://github.com/apache/logging-log4net/search?p=2&q=Flush&unscoped_q=Flush Don't see any other proper usage of `Flush` method being made.

Reply:
Have also quickly scanned the log4net source-code, and it looks like Flush is not called automatically. It is for one to call manually. I guess it comes from the File-Appender nature where closing the file-stream implicit performs flush, so no flush is needed in log4net. But I can see the Flush-method in this library is forever until success, which is not a good behavior when trying to close/shutdown.

Reply:
Think implementing Flush in OnClose for log4net-appender should be a different PR.

Reply:
Sure, not an issue. We can definitely do this in a different PR.